vm_insnhelper.c

```diff
@@ -1438,209 +1440,198 @@ vm_expandarray(VALUE *sp, VALUE ary, rb_num_t num, int flag)
- *   `cc->aux` might be invalid.  Also the found class serial is expected to
- *   hit next time.  In this case we reorder the array and wipe out `cc->aux`.
- *   This is done in the #2 loop below.
- *
- *   ```
- *   from:  +--------------+-----+-----+-----+----+-----+------+-----+
- *          | method_state | (x) | (y) | (z) | me | def | call | aux |
- *          +--------------+-----+-----+-----+----+-----+------+-----+
- *                             \     \    |
- *                              \     \   |
- *                            +- \ --- \ -+
- *                            |   \     \
- *                            |    \     \
- *                            v     v     v
- *   to:    +--------------+-----+-----+-----+----+-----+------+-----+
- *          | method_state | (z) | (x) | (y) | me | def | call | 000 |
- *          +--------------+-----+-----+-----+----+-----+------+-----+
- *                                                               ^^^
- *                                                            wipe out
- *   ```
- *
- */
-static inline bool
-vm_cache_check_for_class_serial(struct rb_call_cache *cc, rb_serial_t class_serial)
-{
-    int i;
-    rb_serial_t j;
+    else {
+        cc_tbl = RCLASS_CC_TBL(klass) = rb_id_table_create(2);
+    }
 
-    /* This is the loop #1 in above description. */
-    for (i = 0; i < numberof(cc->class_serial); i++) {
-        j = cc->class_serial[i];
+    const rb_callable_method_entry_t *cme = rb_callable_method_entry(klass, mid);
 
-        if (! j) {
-            break;
-        }
-        else if (j != class_serial) {
-            continue;
-        }
-        else if (! i) {
-            return true;
-        }
-        else {
-            goto hit;
+    if (cme == NULL) {
+        // undef or not found: can't cache the information
+        VM_ASSERT(vm_cc_cme(vm_cc_empty()) == NULL);
+        return vm_cc_empty();
+    }
+    else {
+        const struct rb_callcache *cc = vm_cc_new(klass, cme, vm_call_general);
+        METHOD_ENTRY_CACHED_SET((struct rb_callable_method_entry_struct *)cme);
+
+        if (ccs == NULL) {
+            VM_ASSERT(cc_tbl != NULL);
+            ccs = vm_ccs_create(klass, cme);
+            rb_id_table_insert(cc_tbl, mid, (VALUE)ccs);
         }
+        vm_ccs_push(klass, ccs, ci, cc);
+
+        VM_ASSERT(vm_cc_cme(cc) != NULL);
+        VM_ASSERT(cme->called_id == mid);
+        VM_ASSERT(vm_cc_cme(cc)->called_id == mid);
+        return cc;
     }
+}
 
-    RB_DEBUG_COUNTER_INC(mc_class_serial_miss);
-    return false;
+MJIT_FUNC_EXPORTED void
+rb_vm_search_method_slowpath(VALUE cd_owner, struct rb_call_data *cd, VALUE klass)
+{
+    const struct rb_callcache *cc = vm_search_cc(klass, cd->ci);
 
-  hit:
-    /* This is the loop #2 in above description. */
-    for (; i > 0; i--) {
-        cc->class_serial[i] = cc->class_serial[i - 1];
+    if (cd_owner) {
+        RB_OBJ_WRITE(cd_owner, &cd->cc, cc);
```

Comment:
Multiple guilds can write this field at once, right?  I guess that is what you call the "atomicity", and you think that works.  Am I correct?

Reply:
a guild can consistent CC so there is no problem.